### PR TITLE
fix(mock): detect $ref in single-element allOf/oneOf/anyOf array items

### PR DESCRIPTION
**Pull request by jakiestfu**
## Summary

- Adds `extractItemsRef` helper to detect `$ref` whether direct on `item.items` or wrapped in a single-element `allOf`/`oneOf`/`anyOf` composition
- Fixes recursion guard so self-referential array schemas emit `[]` instead of `undefined[]`
- Normalizes wrapped items to `{ $ref }` before `resolveMockValue`, preventing enum arrays from double-wrapping as `SomeEnum[][]`
- Only single-element compositions are unwrapped; multi-element compositions still flow through the combine path

### Self-referential schema producing `undefined[]`

```yaml
components:
  schemas:
    Category:
      type: object
      properties:
        id:
          type: integer
          format: int64
        name:
          type: string
        subcategories:
          type: array
          items:
            allOf:
              - $ref: '#/components/schemas/Category'
```

**Before fix:** The recursion guard didn't fire because it only checked `item.items.$ref` (which is `undefined` when wrapped in `allOf`). The code recursed into `combineSchemasMock`, which eventually skipped the already-visited ref but left the parent array case generating `Array.from(...).map(() => undefined)` — producing `undefined[]`.

**After fix:** `extractItemsRef` unwraps the single-element `allOf` and the recursion guard correctly returns `[]`.

### Enum array double-wrapping as `PetStatus[][]`

```yaml
components:
  schemas:
    PetStatus:
      type: string
      enum: [available, pending, sold]
    Pet:
      type: object
      properties:
        name:
          type: string
        previousStatuses:
          type: array
          items:
            allOf:
              - $ref: '#/components/schemas/PetStatus'
```

**Before fix:** The `allOf`-wrapped items went through the composition path, so the outer array case never received the `enums` flag back from `resolveMockValue`. It wrapped the already-correct `faker.helpers.arrayElements(Object.values(PetStatus))` (which itself returns `PetStatus[]`) in an extra `Array.from(...).map(...)`, yielding `PetStatus[][]`.

**After fix:** The wrapped items are normalized to `{ $ref: '#/components/schemas/PetStatus' }` before calling `resolveMockValue`, which resolves the enum through the normal `$ref` path, properly propagates the `enums` flag, and the array case returns the value directly without double-wrapping.

## Test plan

- [x] Unit tests added for recursion guard with direct `$ref`, `allOf`, `oneOf`, `anyOf` wrappers
- [x] Negative test: multi-element `allOf` is not unwrapped
- [x] All existing tests pass (`npx vitest run packages/mock/src/faker/getters/scalar.test.ts` — 40 tests)

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved mock generation for array schemas with circular references. Now detects references even when wrapped by single-element composed schemas (allOf/oneOf/anyOf), returns an empty array for circular cases, and avoids incorrect short-circuits that could produce wrong mocks.

* **Tests**
  * Added tests covering array reference extraction and recursion prevention, including edge cases with nested composed schemas.

[Review Change Stack](https://app.coderabbit.ai/change-stack/orval-labs/orval/pull/3421?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

File: packages/mock/src/faker/getters/scalar.ts [hunk 1]
```diff
@@ -254,22 +254,31 @@ export function getMockScalar({
...
         return { value: '[]', imports: [], name: item.name };
       }
 
+      // If `items` is a single-element `allOf`/`oneOf`/`anyOf` wrapping a
+      // `$ref`, treat it as a direct `$ref`. This avoids double-wrapping when
+      // the inner schema is an enum array (whose `getEnum` already emits
+      // `faker.helpers.arrayElements(...)`) and keeps recursion semantics in
+      // line with direct-$ref items.
+      const resolvedItems =
+        itemsRef && !('$ref' in item.items) ? { $ref: itemsRef } : item.items;
```
[the remark this answers is not in the document]
Comment (jakiestfu) [reply packages/mock/src/faker/getters/scalar.ts:274]: This is intentional and correct for the mock generation context. When `resolveMockValue` encounters a `$ref` schema, it calls `isReference(schema)` → resolves the ref by walking `context.spec` to the target → builds a new schema entirely from the target. **Sibling keywords on a `$ref` object are already ignored** by the existing `resolveMockValue` implementation (this is the standard JSON Reference behavior — `$ref` replaces the entire object).

So even if we preserved siblings like `nullable` by spreading `item.items`, they would be discarded the moment `resolveMockValue` enters its `isReference` branch. The direct-`$ref` case (`item.items: { $ref: '...' }`) has this same behavior — any sibling keywords are dropped.

The second "analogous occurrence" mentioned does not exist in this file — there is only one `resolvedItems` assignment.

[the remark this answers is not in the document]
Comment (jakiestfu) [reply packages/mock/src/faker/getters/scalar.ts:274]: Good robot

**Comment by jakiestfu** [comment]
@melloware I think this is one of the last bugs that is preventing us at Turo from using the faker mocks. We might be interested in exploring support for generating mocks for models too, not just operations, that can come later.

**Comment by melloware** [comment]
Awesome I will check it tomorrow!

**Comment by melloware** [comment]
@jakiestfu did you want to filter the issues and look for label=mock to see if this fixes any open issues or if you want to fix any of those open issues before the next release??

**Comment by jakiestfu** [comment]
@melloware I will look to see if this solves any open issues and follow up with them.

Unfortunately, I likely will not be able to invest much time in tackling general orval issues unless they directly impact our ability to use the library, apologies.

**Review by melloware — APPROVED** [review]

**Comment by melloware** [comment]
8.12.3 is published

**Comment by jakiestfu** [comment]
> 8.12.3 is published

So fast, @melloware, thank you so much!
